### PR TITLE
More helpful type guards

### LINKED ISSUE #13199: no error when no argument in `TypeGuard` function
```py
from typing import TypeGuard

def foo() -> TypeGuard[int]: ... # no error
```
[playground](https://mypy-play.net/?mypy=master&python=3.10&flags=show-error-codes%2Callow-redefinition%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-optional%2Cno-implicit-reexport%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=5c5d885cfc2c3b91a9852459f0806d01)

### PR DESCRIPTION
**Pull request by A5rocks**
A TODO I found while looking at something else got me down this rabbit hole, and I noticed mypy's subpar performance in these two cases.

Short descriptions from *before* I did this, so may be technically inaccurate:
1) check when defining a type guard whether the function is right or not
2) follow kw args if they go to a regular positional first arg 

~~This may fix an issue but I have not conducted any searches.~~
Fixes #13199
Refs #14425

### CODE REVIEW

**Comment by A5rocks** [comment]
mypy primer results seem to be narrowing on self:
```py
    @abstractmethod
    def is_error(self) -> TypeGuard[Error[_TSource, _TError]]:
        """Returns `True` if the result is an `Error` value."""

        raise NotImplementedError
```

This seems... wrong? Firstly, the PEP says this (as mentioned in one of my commit messages):
> If a type guard function is implemented as an instance method or class method, the first positional argument maps to the second parameter (after “self” or “cls”).

This was even **explicitly** rejected! https://peps.python.org/pep-0647/#narrowing-of-implicit-self-and-cls-parameters

Secondly, mypy doesn't even support this!! At the moment:

```py
from typing import TypeGuard

class A:
  def n(self) -> TypeGuard[int]:
    return False

a = A()
if a.n():
  reveal_type(a)  # N: Revealed type is "__main__.A"
```

---

However, I realize I missed `staticmethod` special casing. Will fix.
```py
from typing_extensions import TypeGuard

class Z:
    @staticmethod
    def typeguard(h: object) -> TypeGuard[int]:  # E: TypeGuard functions must have a positional argument
        return isinstance(h, int)

x: object
if Z().typeguard(x):
    reveal_type(x)
if Z.typeguard(x):
    reveal_type(x)
```

File: mypy/checker.py [hunk 1]
```diff
@@ -5284,10 +5284,19 @@ def find_isinstance_check_helper(self, node: Expression) -> tuple[TypeMap, TypeM
...
             elif isinstance(node.callee, RefExpr):
                 if node.callee.type_guard is not None:
-                    # TODO: Follow keyword args or *args, **kwargs
+                    # TODO: Follow *args, **kwargs
                     if node.arg_kinds[0] != nodes.ARG_POS:
-                        self.fail(message_registry.TYPE_GUARD_POS_ARG_REQUIRED, node)
-                        return {}, {}
+                        # the first argument might be used as a kwarg
+                        called_type = get_proper_type(self.lookup_type(node.callee))
+                        assert isinstance(called_type, CallableType)
```
Comment (sobolevn) [inline mypy/checker.py:5291]: I fear that this might also be `OverloadedType` 🤔

File: mypy/checker.py [hunk 2]
```diff
@@ -5284,10 +5284,19 @@ def find_isinstance_check_helper(self, node: Expression) -> tuple[TypeMap, TypeM
...
+                        name = called_type.arg_names[0]
+                        if name in node.arg_names:
+                            idx = node.arg_names.index(name)
+                            # we want the idx-th variable to be narrowed
+                            expr = collapse_walrus(node.args[idx])
```
Comment (sobolevn) [inline mypy/checker.py:5296]: Looks like this part is not tested. There are no tests for a `:=` added.

File: mypy/semanal.py [hunk 3]
```diff
@@ -865,6 +865,20 @@ def analyze_func_def(self, defn: FuncDef) -> None:
                     return
                 assert isinstance(result, ProperType)
                 if isinstance(result, CallableType):
+                    # type guards need to have a positional argument, to spec
```
Comment (sobolevn) [inline mypy/semanal.py:868]: What about pos-only?

**Review by sobolevn — COMMENTED** [review]
Thanks! I have a couple of questions, this is not a full review.

File: mypy/semanal.py (hunk 3 above)
Comment (A5rocks) [reply mypy/semanal.py:868]: I think this works already, will add a test though.

**Comment by A5rocks** [comment]
OK with that tests should be added as requested. I've got one additional question that I've moved over to #14273 because it may cause more debate. Hopefully this as it stands is alright!

File: test-data/unit/check-typeguard.test [hunk 4]
```diff
@@ -639,7 +639,7 @@ if typeguard(n):
     reveal_type(n)
 [builtins fixtures/tuple.pyi]
 [out]
-main:4: error: invalid syntax
+main:4: error: invalid syntax; you likely need to run mypy using Python 3.8 or newer
```
Comment (sobolevn) [inline test-data/unit/check-typeguard.test:642]: You can move these tests to https://github.com/python/mypy/blob/master/test-data/unit/check-python38.test

**Comment by sobolevn** [comment]
Cases that error:

```python
   
    @abstractmethod
    def is_error(self) -> TypeGuard[Error[_TSource, _TError]]:
        """Returns `True` if the result is an `Error` value."""

        raise NotImplementedError

    @abstractmethod
    def is_ok(self) -> TypeGuard[Ok[_TSource, _TError]]:
        """Returns `True` if the result is an `Ok` value."""

        raise NotImplementedError
```

**Comment by A5rocks** [comment]
Yep! As mentioned before, that is invalid as according to the PEP and mypy doesn't support it anyways. Having an error at definition time is likely beneficial.

**Comment by A5rocks** [comment]
Someone ran into this in the issue tracker; I have updated the "fixes" section.

File: test-data/unit/check-typeguard.test [hunk 5]
```diff
@@ -597,3 +597,77 @@ def func(names: Tuple[str, ...]):
...
+
+@overload
+def typeguard(x: object, y: str) -> TypeGuard[str]:
+    ...
+
+@overload
+def typeguard(x: object, y: int) -> TypeGuard[int]:
+    ...
+
+def typeguard(x: object, y: Union[int, str]) -> Union[TypeGuard[int], TypeGuard[str]]:
```
Comment (hauntsaninja) [inline test-data/unit/check-typeguard.test:658]: Not sure this return type makes sense. But can add an error in another PR

**Review by hauntsaninja — APPROVED** [review]
This looks great, thanks!

**Comment by hauntsaninja** [comment]
Oh hmm, the crash on `FasterSpeeding/Tanjun` is new (not to this PR, just new in general)

**Pull request merged**